bumble/hci.py

```diff
@@ -3064,13 +3108,115 @@ class HCI_Read_BD_ADDR_Command(HCI_Command):
+        ('num_supported_standard_codecs', 1),
+        ('standard_codec_id', {'size_field': 'num_supported_standard_codecs'}),
+        ('num_supported_vendor_specific_codecs', 1),
+        (
+            'vendor_specific_codec_id',
+            {
+                'size_field': 'num_supported_vendor_specific_codec',
+                'tranformer': lambda x: x * 4,
+            },
+        ),
+    ]
+)
+class HCI_Read_Local_Supported_Codecs_V1_Command(HCI_Command):
+    '''
+    See Bluetooth spec @ 7.4.8 Read Local Supported Codecs Command
+    '''
+
+
+# -----------------------------------------------------------------------------
+@HCI_Command.command(
+    return_parameters_fields=[
+        ('status', STATUS_SPEC),
+        ('num_supported_standard_codecs', 1),
+        ('standard_codec_id', {'size_field': 'num_supported_standard_codecs'}),
+        ('standard_codec_transport', {'size_field': 'num_supported_standard_codecs'}),
+        ('num_supported_vendor_specific_codecs', 1),
+        (
+            'vendor_specific_codec_id',
+            {
+                'size_field': 'num_supported_vendor_specific_codec',
+                'transformer': lambda x: x * 4,
+            },
+        ),
+        (
+            'vendor_specific_codec_transport',
+            {'size_field': 'num_supported_vendor_specific_codecs'},
+        ),
+    ]
+)
+class HCI_Read_Local_Supported_Codecs_V2_Command(HCI_Command):
     '''
     See Bluetooth spec @ 7.4.8 Read Local Supported Codecs Command
     '''
 
 
+# -----------------------------------------------------------------------------
+@HCI_Command.command(
+    return_parameters_fields=[
+        ('status', STATUS_SPEC),
+        ('simple_pairing_options', 1),
+        ('max_encryption_key_size', 1),
+    ]
+)
+class HCI_Read_Local_Simple_Pairing_Options_Command(HCI_Command):
+    '''
+    See Bluetooth spec @ 7.4.9 Read Local Simple Pairing Options Command
+    '''
+
+
+# -----------------------------------------------------------------------------
+@HCI_Command.command(
+    fields=[
+        ('codec_id', 5),
+        ('logical_transport_type', 1),
+        ('direction', 1),
+    ],
+    return_parameters_fields=[
+        ('status', STATUS_SPEC),
+        ('num_codec_capabilities', 1),
+        ('codec_capability_length', {'size_field': 'num_codec_capabilities'}),
+        (
+            'codec_capability',
+            {
+                'size_field': 'codec_capability_length',
+                'transformer': lambda len_bytes: sum(len_bytes),
+            },
+        ),
+    ],
+)
+class HCI_Read_Local_Supported_Codec_Capabilities_Command(HCI_Command):
+    '''
+    See Bluetooth spec @ 7.4.10 Read Local Supported Codec Capabilities command
+    '''
+
+
+# -----------------------------------------------------------------------------
+@HCI_Command.command(
+    fields=[
+        ('codec_id', 5),
+        ('logical_transport_type', 1),
+        ('direction', 1),
+        ('codec_configuration_length', 1),
```

Comment:
I agree that using this approach does not require passing the length field when constructing the packet.

But this approach does has its disadvantages:
1.  not flexible enough, as it only supports 1-byte length fields, I am not sure whether are multi-byte length fields used at all, though
2. the length field is missing from the explicit packet definition in the code, which sometimes makes it confusing for those who are not familiar with the tricks handled by the proposed spec, and when we debug by inspected packet dumps. Maybe having a explicitly defined field makes it less confusing?

Reply:
So far, there has only been 1-byte item count fields (which makes sense, since HCI packets are never very large, it would be unusual to have more than 256 entries in an array). If this happens at some point, I think it would be easy enough to add a marker, along with the array, to indicate that the item count should be encoded on more than 1 byte.
For the length field, it's true that it won't appear as a property of the object after parsing, but I think that's good. It is always error-prone to have the same information represented in two different ways: in this case, the array fields are arrays of elements, so that includes the "count" information already. Also, when invoking the constructor is much more natural if you pass arrays but don't have to also pass the length of the arrays (ex: `HCI_Foo(foos=[1,2,3], blas=[4,5,6])` rather than `HCI_Foo(entry_count=3, foos=[1,2,3], blas=[4,5,6])`. 
Also, that's already the way all the existing HCI objects with array fields work (ex: `HCI_LE_Set_Extended_Scan_Parameters_Command` and others, see the tests for examples of how the constructors are invoked), so we want to keep that model.

Reply:
Another frequent setting in my tests requires non-conformant tests, e.g., to test code running on remote devices with a length field that is not consistent with the number of bytes present. With this approach, construction of such packets is completely impossible.

This approach does provide the convenience, but loses flexibility.

Reply:
Creating non-compliant HCI packets is a special case that's probably best supported by other means than using the normal object constructor (there are already a lot of fields that are serialized automatically and that you can't make an invalid encoding for by just using the constructor, like address fields, or the size of fields, etc.). If you know exactly what you want to modify in a packet to make it non-compliant in some way, the best is to either create a good one and change one of its bytes, or just create one with a pre-serialized byte array.